### PR TITLE
Adjust apiserver-proxy to allow k8s apiserver to be secured

### PR DESCRIPTION
**Pull request by piotrmsc**
**Description**
As with the decision to adjust kube-rbac-proxy with our needs for securing k8s apiserver, this PR serves all required code changes on top of kube-rbac-proxy.
Changes proposed in this pull request:

- Code changes in the apiserver-proxy which allows k8s apiserver to be secured

**Related issue(s)**
See also #707 >

### CODE REVIEW

File: components/apiserver-proxy/internal/proxy/proxy.go [hunk 1]
```diff
@@ -9,13 +9,18 @@ import (
...
 	yaml "gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/authorization/authorizer"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 	clientset "k8s.io/client-go/kubernetes"
 )
 
+const KUBERNETES_SERVICE = "kubernetes.default"
```
Comment (sjanota) [inline components/apiserver-proxy/internal/proxy/proxy.go]: This constant seems to be unused. If so, please remove it.

File: components/apiserver-proxy/internal/proxy/proxy.go [hunk 2]
```diff
@@ -201,3 +225,15 @@ func (c *Config) DeepCopy() *Config {
 
 	return res
 }
+
+func newRequestInfoResolver() *apirequest.RequestInfoFactory {
+	apiPrefixes := sets.NewString(strings.Trim("/apis", "/"))
```
Comment (sjanota) [inline components/apiserver-proxy/internal/proxy/proxy.go]: Why not simply:
```
apiPrefixes := sets.NewString("apis", "api")
legacyAPIPrefixes := sets.NewString("api")
```

File: components/apiserver-proxy/internal/proxy/proxy.go (hunk 1 above)
Comment (piotrmsc) [reply components/apiserver-proxy/internal/proxy/proxy.go]: It is used in main.go

File: components/apiserver-proxy/internal/proxy/proxy.go (hunk 2 above)
Comment (piotrmsc) [reply components/apiserver-proxy/internal/proxy/proxy.go]: yes, it was simplified

File: components/apiserver-proxy/cmd/proxy/main.go [hunk 3]
```diff
@@ -162,21 +146,32 @@ func main() {
...
 	if err != nil {
 		glog.Fatalf("Failed to create rbac-proxy: %v", err)
 	}
 
-	proxy := httputil.NewSingleHostReverseProxy(upstreamURL)
+	proxyForApiserver := strings.Contains(cfg.upstream, proxy.KUBERNETES_SERVICE)
+
+	rp := httputil.NewSingleHostReverseProxy(upstreamURL)
+	
+	if proxyForApiserver {
```
Comment (sjanota) [inline components/apiserver-proxy/cmd/proxy/main.go]: Are we going to use it in ant other way then as proxy for apiserver? If not I'd refrain from using conditions here and just hardcode this one case.

File: components/apiserver-proxy/cmd/proxy/main.go [hunk 4]
```diff
@@ -221,11 +216,11 @@ func main() {
 	}
 
 	if cfg.insecureListenAddress != "" {
-		if cfg.upstreamForceH2C {
+		if cfg.upstreamForceH2C && !proxyForApiserver {
```
Comment (sjanota) [inline components/apiserver-proxy/cmd/proxy/main.go]: Same here, if we are going to use it only for apiserver this whole if only clutters

File: components/apiserver-proxy/cmd/proxy/main.go [hunk 5]
```diff
@@ -291,9 +286,10 @@ func main() {
 func initKubeConfig(kcLocation string) *rest.Config {
 
 	if kcLocation != "" {
+		println(kcLocation)
```
Comment (sjanota) [inline components/apiserver-proxy/cmd/proxy/main.go]: Forgotten debug statement?

File: components/apiserver-proxy/cmd/proxy/main.go (hunk 3 above)
Comment (piotrmsc) [reply components/apiserver-proxy/cmd/proxy/main.go]: yes, it should be possible to use it for another app as well although here it is designed for apisever :)

File: components/apiserver-proxy/cmd/proxy/main.go (hunk 4 above)
Comment (piotrmsc) [reply components/apiserver-proxy/cmd/proxy/main.go]: like above

**Pull request merged**